frontend/src/scenes/insights/views/LineGraph/LineGraph.tsx

```diff
@@ -537,20 +535,50 @@ export function LineGraph_({
                         const bounds = canvasRef.current.getBoundingClientRect()
-                        const horizontalBarTopOffset = isHorizontal ? tooltip.caretY - tooltipEl.clientHeight / 2 : 0
-                        const tooltipClientTop = bounds.top + window.pageYOffset + horizontalBarTopOffset
-
-                        const chartClientLeft = bounds.left + window.pageXOffset
-                        const defaultOffsetLeft = Math.max(chartClientLeft, chartClientLeft + tooltip.caretX + 8)
-                        const maxXPosition = bounds.right - tooltipEl.clientWidth
-                        const tooltipClientLeft =
-                            defaultOffsetLeft > maxXPosition
-                                ? chartClientLeft + tooltip.caretX - tooltipEl.clientWidth - 8 // If tooltip is too large (or close to the edge), show it to the left of the data point instead
-                                : defaultOffsetLeft
-
-                        tooltipEl.style.top = Math.min(tooltipClientTop, window.innerHeight) + 'px'
-                        tooltipEl.style.left = Math.min(tooltipClientLeft, window.innerWidth) + 'px'
+
+                        const setPositionAndMakeVisible = (): void => {
+                            const tooltipHeight = tooltipEl.clientHeight // this may be stale, see comment below
+
+                            const horizontalBarTopOffset = isHorizontal ? tooltip.caretY - tooltipHeight / 2 : 0
+                            let tooltipClientTop = bounds.top + windowPageYOffset + horizontalBarTopOffset
+
+                            const chartClientLeft = bounds.left + windowPageXOffset
+                            const defaultOffsetLeft = Math.max(chartClientLeft, chartClientLeft + tooltip.caretX + 8)
+                            const maxXPosition = bounds.right - tooltipEl.clientWidth
+                            const tooltipClientLeft =
+                                defaultOffsetLeft > maxXPosition
+                                    ? chartClientLeft + tooltip.caretX - tooltipEl.clientWidth - 8 // If tooltip is too large (or close to the edge), show it to the left of the data point instead
+                                    : defaultOffsetLeft
+
+                            // prevent the bottom of the tooltip from being below the bottom of the viewport
+                            const maxYPosition = windowPageYOffset + windowInnerHeight - tooltipHeight
+                            tooltipClientTop = Math.min(tooltipClientTop, maxYPosition)
+
+                            tooltipEl.style.top = tooltipClientTop + 'px'
+                            tooltipEl.style.left = tooltipClientLeft + 'px'
+
+                            tooltipEl.style.opacity = '1'
+                            tooltipEl.style.display = 'initial'
+                        }
+
+                        setPositionAndMakeVisible()
+                        setTimeout(setPositionAndMakeVisible, 0)
```

Comment:
I assume this is still a draft because of the dangling setTimeout ;) 

This would be a big no-no as it would never end

Reply:
The setTimeout is used for the second rendering pass I think:
> // Do the work in a function which is called both immediately and after a delay.

Not sure if I understand this, did you maybe read `setInterval`?
> This would be a big no-no as it would never end

Reply:
Ah wait yeah I misread and thought it was called from within the function itself to create a setInterval-like thing.

Still think that a comment is needed for why that is there as it was not super obvious to me